File: lib/product/semantic_diff.ml
type path_change = {
  source : string;
  sink : string;
  path : string list;
  effect_name : string;
}

type change =
  | New_reachable_path of path_change
  | Capability_added of Ir.capability
  | Capability_removed of Ir.capability
  | Dependency_became_mutable of string
  | Property_changed of { property : string; before : string; after : string }

type t = {
  schema : string;
  base_digest : string;
  head_digest : string;
  changes : change list;
}

let node_value node =
  List.fold_left
    (fun accumulator (_, value) -> Abstract_value.join accumulator value)
    Abstract_value.bottom node.Ir.attributes

let effects (node : Ir.node) =
  let inferred =
    if node.Ir.kind = Ir.Command then
      (Script_adapter.analyze Bash node.name).effects
    else []
  in
  Util.deduplicate_compare Stdlib.compare (node.effects @ inferred)

let attack_paths graph =
  let sources =
    List.filter
      (fun (node : Ir.node) -> Abstract_value.is_untrusted (node_value node))
      graph.Ir.nodes
  and sinks =
    graph.nodes
    |> List.concat_map (fun node ->
        effects node |> List.map (fun observable -> (node, observable)))
  in
  List.concat_map
    (fun (source : Ir.node) ->
      sinks
      |> List.filter_map (fun ((sink : Ir.node), observable) ->
          Graph_algorithms.shortest_path
            ~edge_kinds:
              [
                Ir.Data; Ir.Read; Ir.Write; Ir.Persist; Ir.Call_edge; Ir.Control;
              ]
            graph source.id sink.id
          |> Option.map (fun path ->
              {
                source = source.id;
                sink = sink.id;
                path = List.map (fun (node : Ir.node) -> node.id) path;
                effect_name = Ir.effect_name observable;
              })))
    sources

let path_key path =
  String.concat "\000"
    (path.source :: path.sink :: path.effect_name :: path.path)

let capabilities graph =
  graph.Ir.nodes
  |> List.concat_map (fun (node : Ir.node) -> node.capabilities)
  |> Util.deduplicate_compare Stdlib.compare

let mutable_reference reference =
  match String.rindex_opt reference '@' with
  | None -> false
  | Some index ->
      let revision =
        String.sub reference (index + 1) (String.length reference - index - 1)
      in
      String.length revision < 40
      && not (Util.starts_with ~prefix:"sha256:" revision)

let mutable_calls graph =
  graph.Ir.nodes
  |> List.filter_map (fun (node : Ir.node) ->
      if node.kind = Ir.Call && mutable_reference node.name then Some node.name
      else None)
  |> Util.deduplicate_strings

let compare_with_properties base head base_properties head_properties =
  let base_paths = attack_paths base and head_paths = attack_paths head in
  let new_paths =
    List.filter
      (fun path ->
        not (List.exists (fun old -> path_key old = path_key path) base_paths))
      head_paths
    |> List.map (fun value -> New_reachable_path value)
  in
  let base_caps = capabilities base and head_caps = capabilities head in
  let added =
    List.filter (fun value -> not (List.mem value base_caps)) head_caps
    |> List.map (fun value -> Capability_added value)
  and removed =
    List.filter (fun value -> not (List.mem value head_caps)) base_caps
    |> List.map (fun value -> Capability_removed value)
  and mutable_added =
    let old = mutable_calls base in
    mutable_calls head
    |> List.filter (fun value -> not (List.mem value old))
    |> List.map (fun value -> Dependency_became_mutable value)
  and property_changes =
    head_properties
    |> List.filter_map (fun (property : Property.t) ->
        match
          List.find_opt
            (fun (candidate : Property.t) -> candidate.id = property.id)
            base_properties
        with
        | None ->
            Some
              (Property_changed
                 {
                   property = property.id;
                   before = "NotApplicable";
                   after = Property.state_name property.state;
                 })
        | Some before ->
            let before_name = Property.state_name before.state
            and after_name = Property.state_name property.state in
            if before_name = after_name then None
            else
              Some
                (Property_changed
                   {
                     property = property.id;
                     before = before_name;
                     after = after_name;
                   }))
    |> List.sort (fun left right ->
        match (left, right) with
        | Property_changed left, Property_changed right ->
            String.compare left.property right.property
        | _ -> 0)
  in
  {
    schema = "semantic-diff-v1";
    base_digest =
      "sha256:" ^ Sha256.digest_string (Json.to_string (Ir.to_json base));
    head_digest =
      "sha256:" ^ Sha256.digest_string (Json.to_string (Ir.to_json head));
    changes = new_paths @ added @ removed @ mutable_added @ property_changes;
  }

let compare base head = compare_with_properties base head [] []

let compare_program base head =
  let base_program = Program_graph.compose base
  and head_program = Program_graph.compose head
  and base_verification = Verifier.verify_program ~persona:Verifier.Audit base
  and head_verification =
    Verifier.verify_program ~persona:Verifier.Audit head
  in
  compare_with_properties base_program head_program base_verification.properties
    head_verification.properties

let change_json = function
  | New_reachable_path path ->
      Json.Object
        [
          ("effect", Json.String path.effect_name);
          ("kind", Json.String "new_reachable_path");
          ( "path",
            Json.Array (List.map (fun value -> Json.String value) path.path) );
          ("sink", Json.String path.sink);
          ("source", Json.String path.source);
        ]
  | Capability_added value ->
      Json.Object
        [
          ("capability", Json.String (Ir.capability_name value));
          ("kind", Json.String "capability_added");
        ]
  | Capability_removed value ->
      Json.Object
        [
          ("capability", Json.String (Ir.capability_name value));
          ("kind", Json.String "capability_removed");
        ]
  | Dependency_became_mutable value ->
      Json.Object
        [
          ("kind", Json.String "dependency_became_mutable");
          ("reference", Json.String value);
        ]
  | Property_changed { property; before; after } ->
      Json.Object
        [
          ("after", Json.String after);
          ("before", Json.String before);
          ("kind", Json.String "property_changed");
          ("property", Json.String property);
        ]

let to_json difference =
  Json.Object
    [
      ("base_digest", Json.String difference.base_digest);
      ("changes", Json.Array (List.map change_json difference.changes));
      ("head_digest", Json.String difference.head_digest);
      ("schema", Json.String difference.schema);
    ]

let to_canonical_json difference = Json.to_string (to_json difference) ^ "\n"

File: crates/product/src/semantic_diff.rs
use std::collections::{BTreeMap, BTreeSet, VecDeque};
use workflow_verifier_domain::{Capability, EdgeKind, Graph, ObservableEffect};
use workflow_verifier_foundation::{DependencyClass, JsonValue, classify_reference};

#[derive(Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
pub struct PathChange {
    pub source: String,
    pub sink: String,
    pub path: Vec<String>,
    pub effect: ObservableEffect,
}

#[derive(Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
pub enum SemanticChange {
    NewReachablePath(PathChange),
    CapabilityAdded(Capability),
    CapabilityRemoved(Capability),
    DependencyBecameMutable(String),
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct SemanticDiff {
    pub base_digest: String,
    pub head_digest: String,
    pub changes: Vec<SemanticChange>,
}

impl SemanticDiff {
    fn to_json(&self) -> JsonValue {
        JsonValue::Object(BTreeMap::from([
            (
                "base_digest".to_owned(),
                JsonValue::String(self.base_digest.clone()),
            ),
            (
                "changes".to_owned(),
                JsonValue::Array(self.changes.iter().map(change_json).collect()),
            ),
            (
                "head_digest".to_owned(),
                JsonValue::String(self.head_digest.clone()),
            ),
            (
                "schema".to_owned(),
                JsonValue::String("semantic-diff-v1".to_owned()),
            ),
        ]))
    }

    #[must_use]
    pub fn to_canonical_json(&self) -> String {
        self.to_json().canonical_line()
    }

    #[must_use]
    pub fn verify_digests(&self, base: &Graph, head: &Graph) -> bool {
        self.base_digest == graph_digest(base) && self.head_digest == graph_digest(head)
    }
}

fn graph_digest(graph: &Graph) -> String {
    graph.to_json().canonical_digest()
}

fn capabilities(graph: &Graph) -> BTreeSet<Capability> {
    graph
        .nodes
        .iter()
        .flat_map(|node| node.capabilities.iter().copied())
        .collect()
}

fn mutable_calls(graph: &Graph) -> BTreeSet<String> {
    graph
        .nodes
        .iter()
        .filter(|node| {
            node.kind == workflow_verifier_domain::NodeKind::Call
                && classify_reference(&node.name) == DependencyClass::Mutable
        })
        .map(|node| node.name.clone())
        .collect()
}

fn attack_paths(graph: &Graph) -> BTreeSet<PathChange> {
    let sources: Vec<_> = graph
        .nodes
        .iter()
        .filter(|node| {
            node.attributes
                .values()
                .any(workflow_verifier_domain::AbstractValue::is_untrusted)
        })
        .collect();
    let sinks: Vec<_> = graph
        .nodes
        .iter()
        .flat_map(|node| node.effects.iter().map(move |effect| (node, *effect)))
        .collect();
    let mut paths = BTreeSet::new();
    for source in sources {
        for (sink, effect) in &sinks {
            if let Some(path) = shortest_path(graph, &source.id, &sink.id) {
                paths.insert(PathChange {
                    source: source.id.clone(),
                    sink: sink.id.clone(),
                    path,
                    effect: *effect,
                });
            }
        }
    }
    paths
}

fn shortest_path(graph: &Graph, start: &str, goal: &str) -> Option<Vec<String>> {
    let kinds = [
        EdgeKind::Data,
        EdgeKind::Read,
        EdgeKind::Write,
        EdgeKind::Persist,
        EdgeKind::Call,
        EdgeKind::Control,
    ];
    let mut queue = VecDeque::from([start.to_owned()]);
    let mut previous: BTreeMap<String, Option<String>> = BTreeMap::from([(start.to_owned(), None)]);
    while let Some(current) = queue.pop_front() {
        if current == goal {
            let mut path = Vec::new();
            let mut cursor = Some(current);
            while let Some(id) = cursor {
                cursor = previous.get(&id).cloned().flatten();
                path.push(id);
            }
            path.reverse();
            return Some(path);
        }
        let mut successors: Vec<_> = graph
            .edges
            .iter()
            .filter(|edge| edge.from == current && kinds.contains(&edge.kind))
            .map(|edge| edge.to.clone())
            .collect();
        successors.sort();
        successors.dedup();
        for successor in successors {
            if !previous.contains_key(&successor) {
                previous.insert(successor.clone(), Some(current.clone()));
                queue.push_back(successor);
            }
        }
    }
    None
}

#[must_use]
pub fn semantic_diff(base: &Graph, head: &Graph) -> SemanticDiff {
    let base_paths = attack_paths(base);
    let head_paths = attack_paths(head);
    let base_caps = capabilities(base);
    let head_caps = capabilities(head);
    let base_mutable = mutable_calls(base);
    let head_mutable = mutable_calls(head);
    let mut changes = Vec::new();
    changes.extend(
        head_paths
            .difference(&base_paths)
            .cloned()
            .map(SemanticChange::NewReachablePath),
    );
    changes.extend(
        head_caps
            .difference(&base_caps)
            .copied()
            .map(SemanticChange::CapabilityAdded),
    );
    changes.extend(
        base_caps
            .difference(&head_caps)
            .copied()
            .map(SemanticChange::CapabilityRemoved),
    );
    changes.extend(
        head_mutable
            .difference(&base_mutable)
            .cloned()
            .map(SemanticChange::DependencyBecameMutable),
    );
    SemanticDiff {
        base_digest: graph_digest(base),
        head_digest: graph_digest(head),
        changes,
    }
}

fn change_json(change: &SemanticChange) -> JsonValue {
    match change {
        SemanticChange::NewReachablePath(path) => JsonValue::Object(BTreeMap::from([
            (
                "effect".to_owned(),
                JsonValue::String(path.effect.name().to_owned()),
            ),
            (
                "kind".to_owned(),
                JsonValue::String("new_reachable_path".to_owned()),
            ),
            (
                "path".to_owned(),
                JsonValue::Array(path.path.iter().cloned().map(JsonValue::String).collect()),
            ),
            ("sink".to_owned(), JsonValue::String(path.sink.clone())),
            ("source".to_owned(), JsonValue::String(path.source.clone())),
        ])),
        SemanticChange::CapabilityAdded(value) => JsonValue::Object(BTreeMap::from([
            (
                "capability".to_owned(),
                JsonValue::String(value.name().to_owned()),
            ),
            (
                "kind".to_owned(),
                JsonValue::String("capability_added".to_owned()),
            ),
        ])),
        SemanticChange::CapabilityRemoved(value) => JsonValue::Object(BTreeMap::from([
            (
                "capability".to_owned(),
                JsonValue::String(value.name().to_owned()),
            ),
            (
                "kind".to_owned(),
                JsonValue::String("capability_removed".to_owned()),
            ),
        ])),
        SemanticChange::DependencyBecameMutable(reference) => JsonValue::Object(BTreeMap::from([
            (
                "kind".to_owned(),
                JsonValue::String("dependency_became_mutable".to_owned()),
            ),
            ("reference".to_owned(), JsonValue::String(reference.clone())),
        ])),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::collections::BTreeMap;
    use workflow_verifier_domain::{
        AbstractValue, Condition, Edge, Node, NodeKind, Phase, Provider, Secrecy, Trust,
    };
    use workflow_verifier_foundation::Span;

    fn node(kind: NodeKind, name: &str) -> Node {
        Node::simple(Provider::Github, kind, name, Phase::Run, Span::default())
    }

    fn path_graph(with_path: bool) -> (Graph, Node, Node) {
        let mut source = node(NodeKind::Resource, "untrusted source");
        source.attributes = BTreeMap::from([(
            "value".to_owned(),
            AbstractValue::string_constant("input", Trust::Untrusted, Secrecy::Public, Vec::new()),
        )]);
        let middle = node(NodeKind::Step, "middle");
        let mut sink = node(NodeKind::Effect, "deployment");
        sink.effects = vec![ObservableEffect::DeploymentChange];
        let mut graph = Graph::empty(Provider::Github, "workflow.yml");
        graph.add_entrypoint(source.id.clone());
        for item in [&source, &middle, &sink] {
            graph.add_node(item.clone());
        }
        if with_path {
            graph.add_edge(Edge::simple(
                EdgeKind::Data,
                source.id.clone(),
                middle.id.clone(),
            ));
            graph.add_edge(Edge::simple(
                EdgeKind::Control,
                middle.id.clone(),
                sink.id.clone(),
            ));
        } else {
            graph.add_edge(Edge::new(
                EdgeKind::Grant,
                source.id.clone(),
                sink.id.clone(),
                Condition::True,
                None,
            ));
        }
        graph.finalize();
        (graph, source, sink)
    }

    #[test]
    fn attack_path_diff_uses_only_semantic_edges_and_preserves_exact_shortest_path() {
        let (base, _, _) = path_graph(false);
        let (head, source, sink) = path_graph(true);
        assert_eq!(shortest_path(&base, &source.id, &sink.id), None);
        assert_eq!(
            shortest_path(&head, &source.id, &sink.id),
            Some(vec![
                source.id.clone(),
                head.nodes
                    .iter()
                    .find(|node| node.name == "middle")
                    .expect("middle node")
                    .id
                    .clone(),
                sink.id.clone(),
            ])
        );
        assert_eq!(
            shortest_path(&head, &source.id, &source.id),
            Some(vec![source.id.clone()])
        );
        assert_eq!(attack_paths(&base), BTreeSet::new());
        let paths = attack_paths(&head);
        assert_eq!(paths.len(), 1);
        let path = paths.first().expect("attack path");
        assert_eq!(path.source, source.id);
        assert_eq!(path.sink, sink.id);
        assert_eq!(path.effect, ObservableEffect::DeploymentChange);

        let difference = semantic_diff(&base, &head);
        assert!(
            difference
                .changes
                .contains(&SemanticChange::NewReachablePath(path.clone()))
        );
        assert!(difference.verify_digests(&base, &head));
        assert!(!difference.verify_digests(&head, &head));
        assert!(!difference.verify_digests(&base, &base));
        assert!(
            difference
                .to_canonical_json()
                .contains("new_reachable_path")
        );
    }

    #[test]
    fn capability_and_mutability_changes_require_the_right_node_shape() {
        let mut base = Graph::empty(Provider::Github, "base.yml");
        let mut removed = node(NodeKind::Step, "removed capability");
        removed.capabilities = vec![Capability::RepositoryRead];
        base.add_node(removed);
        let mut head = Graph::empty(Provider::Github, "head.yml");
        let mut added = node(NodeKind::Step, "added capability");
        added.capabilities = vec![Capability::RepositoryWrite];
        let mutable_call = node(NodeKind::Call, "owner/action@main");
        let mutable_non_call = node(NodeKind::Step, "owner/not-a-call@main");
        let immutable_call = node(NodeKind::Call, &format!("owner/action@{}", "a".repeat(40)));
        for item in [
            added,
            mutable_call.clone(),
            mutable_non_call,
            immutable_call,
        ] {
            head.add_node(item);
        }
        assert_eq!(
            mutable_calls(&head),
            BTreeSet::from([mutable_call.name.clone()])
        );
        let difference = semantic_diff(&base, &head);
        assert!(
            difference
                .changes
                .contains(&SemanticChange::CapabilityAdded(
                    Capability::RepositoryWrite
                ))
        );
        assert!(
            difference
                .changes
                .contains(&SemanticChange::CapabilityRemoved(
                    Capability::RepositoryRead
                ))
        );
        assert!(
            difference
                .changes
                .contains(&SemanticChange::DependencyBecameMutable(mutable_call.name))
        );
    }
}
